E2E: add spec for SMS OTP authentication.

worldomonation opened the pull request:
#### Proposed Changes

This PR adds the the SMS OTP authentication spec alongside a host of changes.

Note, that while this spec _should_ be stable, restrictions imposed on SMS OTP (1-minute interval between requests) means that possibility of the test flaking out is not zero.

Key changes:
- revised behavior of `EmailClient`; while the name remains EmailClient, its method names and JSDoc has been revamped to remove references to email where possible.
- revised behavior of `TestAccount`; to add both SMS OTP and TOTP support, the `TestAccounts.authenticate` method and its downstream components are updated to branch off.
- update typing of Secrets.secrets.testAccounts.

Other changes:
- upgrade `mailosaur` dependency to 8.4.0;
- add unit tests.

#### Testing Instructions

Ensure the following:
  - [x] Gutenberg E2E (desktop)
  - [x] Gutenberg E2E (mobile)
  - [x] Calypso E2E (desktop)
  - [x] Calypso E2E (mobile)
  - [x] Pre-Release E2E

Related to https://github.com/Automattic/wp-calypso/issues/64841
Project Thread: pciE2j-18A-p2

worldomonation committed:
Changes to EmailClient:
- make it work with SMS as well;
- update method names to be more generic.

Changes to SecretsManager:
- add new fields for SMS OTP support.

Changes to TestAccount:
- rework the OTP verification process to support both SMS and TOTP;
- use the SecretsManager-retrieved credentials directly.

worldomonation committed:
Tests:
- update usage of EmailClient method names and signatures;
- add new spec Authentication: SMS 2FA

worldomonation committed:
Add new user smsUser

worldomonation commented on test/e2e/specs/authentication/authentication__sms.ts:
```diff
@@ -0,0 +1,26 @@
+/**
+ * @group calypso-release
```
One possible problem with this spec is that SMS OTP can only be requested once per minute. 
In addition, anecdotally it appears that any login attempt during this one-minute blackout period resets the timer, though this isn't confirmed.

What this means is that in rare cases where Pre-Release Tests are lined up one after another in the queue, it _may_ result in an unexpected failure of this spec, which would appear to be a flaky test failure to the developer.

It may be necessary to keep a close eye on this test and immediately pull the test from rotation if we find the flakiness exceeds an acceptable level.

lsl replied:
Could you put this comment in the code? Literal copy paste works for me. Just want to explain the calypso-release usage and potential downsides from removing it / changing it. The extra bits about potential reasons this test could one day turn flaky is helpful for debugging too.

dpasque replied:
I just checked the average run of our pre-release tests, and it looks like we're around 2 minutes for the time being, so hopefully we should be good for a while! But +1 to the code comment :)

worldomonation replied:
Thanks for the suggestion, good idea. I'll add what I wrote in the comment above into the docblock.

worldomonation replied:
Addressed in 2f93fdf836bdf423be0769629391e14c14208962.

worldomonation commented on packages/calypso-e2e/src/lib/test-account.ts:
```diff
@@ -81,16 +87,41 @@ export class TestAccount {
+	 * @throws {Error} If TOTP secrets are missing for the user.
 	 */
-	getTOTP(): string | undefined {
+	getTOTP(): string {
```
Throw an error instead of returning an `undefined`.

I personally do not think returning an `undefined` or `null` is useful in situations like this where the TOTP secret doesn't exist yet the code is being asked to generate one.

worldomonation commented on packages/calypso-e2e/src/lib/test-account.ts:
```diff
@@ -55,9 +58,12 @@ export class TestAccount {
-		if ( verificationCode ) {
-			await loginPage.submitVerificationCode( verificationCode );
+		// Handle possible 2FA.
+		if ( this.credentials.totpKey ) {
+			await loginPage.submitVerificationCode( this.getTOTP() );
+		}
+		if ( this.credentials.smsNumber ) {
+			await loginPage.submitVerificationCode( await this.getSMSOTP() );
```
This logic is predicated on the current implementation of 2FA where the active form of 2FA is mutually exclusive; that is, only either SMS or TOTP may be active at any given time.

lsl replied:
Should this be an `else if`? It's semantics and I sort of get why you wouldn't but submitting both codes doesn't make sense either.

worldomonation replied:
You're right - either it should be an `if/else`, or have a `return` statement for each of them. I'll update.

worldomonation replied:
Addressed in 2f93fdf836bdf423be0769629391e14c14208962.

worldomonation commented on packages/calypso-e2e/src/lib/test-account.ts:
```diff
@@ -3,26 +3,29 @@ import path from 'path';
 export class TestAccount {
 	readonly accountName: TestAccountName;
-	readonly credentials: AccountCredentials;
+	readonly credentials: TestAccountCredentials;
```
This is a broadening of the type definition for the variable `credentials`.

`AccountCredentials` narrowed the available fields to just `username, password, totpKey` and was defined separately in `DataHelper`. This was a carryover from the days where the SecretManager did not have type-checking.

Now that SecretsManager has type-checking, we are safe to use the entire returned value from SecretsManager when accessing a specific test account, ie.

```
const credentials = SecretManager.sercrets.testAccount.defaultUser`
```

worldomonation committed:
Update Mailosaur dependency to 8.4.0.
Update EmailClient.get2FACodedFromMessage to use the new structure offered by Mailosaur.

worldomonation committed:
Fix method call in Authentication: Magic Link

worldomonation committed:
Fix issue with SecretsManager types

worldomonation committed:
Add unit tests

worldomonation committed:
Fix WPCC test.

worldomonation committed:
Mock SecretsManager in unit tests

lsl commented on packages/calypso-e2e/src/lib/test-account.ts:
```diff
@@ -3,26 +3,29 @@ import path from 'path';
 import { TOTPClient } from '../totp-client';
 import { LoginPage } from './pages/login-page';
-import type { AccountCredentials } from '../types/data-helper.types';
+// import type { AccountCredentials } from '../types/data-helper.types';
```
```suggestion
```

worldomonation replied:
Ah my bad, thanks for catching that.

worldomonation replied:
Addressed in 2f93fdf836bdf423be0769629391e14c14208962.

lsl reviewed (approved):
Nice work, LGTM

dpasque commented on packages/calypso-e2e/src/test/email-client.test.ts:
```diff
@@ -0,0 +1,90 @@
+	} );
+
+	test( 'Two codes in message', function () {
+		const message = JSON.parse( JSON.stringify( baseSMSMessage ) );
```
100% optional style suggestions!

But I run into this problem when unit testing all the time! 😆 I'll have a base object that you want to tweak throughout tests for different cases, but I need a fresh object reference each time. You can definitely do the stringify and re-parse trick, but my personal preference is to just change `baseSMSMessage` to a function --> `createBaseSMSMessage()` -- then you get a fresh reference every time without having to do any JSON tricks!

Again, totally optional as it's just a style preference thing, but thought I'd share in case it's helpful :)

worldomonation replied:
That's useful and more in line with what we have done elsewhere (including selectors).

worldomonation replied:
Addressed in 2f93fdf836bdf423be0769629391e14c14208962.

dpasque reviewed (approved):
These changes are awesome, nice work! 🚀  Thanks for doing these! 😄

worldomonation committed:
Revise PR based on suggestions:
- add comment about test stability into the docblock of the spec.
- remove commented out import from TestAccount.
- add return statements to the `if` statements in TestAccount.logInViaLoginpage.
- update unit tests for EmailClient by using partial interfaces.

worldomonation commented:
@lsl @dpasque Thank you for the reviews, I really appreciate it 🙇
